README.md
# mxnet-cortex-example

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-04-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-04-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/mxnet-cortex-example/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/mxnet-cortex-example/compare/0.1.0...0.1.1


test/mxnet_cortex_example/core_test.clj
(ns mxnet-cortex-example.core-test
  (:require [clojure.test :refer :all]
            [mxnet-cortex-example.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/mxnet_cortex_example/core.clj
(ns mxnet-cortex-example.core
  (:require
    [org.apache.clojure-mxnet.callback :as callback]
    [org.apache.clojure-mxnet.context :as context]
    [org.apache.clojure-mxnet.dtype :as d]
    [org.apache.clojure-mxnet.eval-metric :as eval-metric]
    [org.apache.clojure-mxnet.executor :as executor]
    [org.apache.clojure-mxnet.initializer :as initializer]
    [org.apache.clojure-mxnet.lr-scheduler :as lr-scheduler]
    [org.apache.clojure-mxnet.io :as mx-io]
    [org.apache.clojure-mxnet.module :as m]
    [org.apache.clojure-mxnet.util :as util]
    [org.apache.clojure-mxnet.ndarray :as ndarray]
    [org.apache.clojure-mxnet.optimizer :as optimizer]
    [org.apache.clojure-mxnet.shape :as mx-shape]
    [org.apache.clojure-mxnet.symbol :as sym]
    [org.apache.clojure-mxnet.visualization :as viz]))

;; Defining the computation graph of the Model
(defn get-symbol []
  (as-> (sym/variable "data") data
    (sym/fully-connected "fc1" {:data data :num-hidden 512})
    (sym/activation "act1" {:data data :act-type "relu"})
    (sym/dropout "drop1" {:data data :p 0.5})
    (sym/fully-connected "fc2" {:data data :num-hidden 128})
    (sym/activation "act2" {:data data :act-type "relu"})
    (sym/dropout "drop2" {:data data :p 0.5})
    (sym/fully-connected "fc3" {:data data :num-hidden 16})
    (sym/activation "act3" {:data data :act-type "relu"})
    (sym/fully-connected "fc4" {:data data :num-hidden 1})
    (sym/linear-regression-output "linear_regression" {:data data})))

;; Model Vizualisation

(defn render-model!
  "Render the `model-sym` and saves it as a png file in `path/model-name.png`
  using graphviz."
  [{:keys [model-name model-sym input-data-shape path]}]
  (let [dot (viz/plot-network
              model-sym
              {"data" input-data-shape}
              {:title model-name
               :node-attrs {:shape "oval" :fixedsize "false"}})]
    (viz/render dot "dot" "png" model-name path)))

;; Creating Datasets as NDArrays

(defn dataset->X-Y
  [dataset]
  (let [n-datapoints (count dataset)
        n-features (-> dataset first :features count)]
    {:X (ndarray/array (->> dataset (map :features) flatten (into []))
                       [n-datapoints n-features])
     :Y (ndarray/array (->> dataset (map :score) flatten (into []))
                       [n-datapoints 1])}))

(defn filename->dataset!
  [filename]
  (-> filename
      slurp
      read-string))

(defonce test-X-Y
  (-> "data/test.txt"
      filename->dataset!
      dataset->X-Y))

(defonce train-X-Y
  (-> "data/train.txt"
      filename->dataset!
      dataset->X-Y))

(def batch-size 2000)

(def test-iter
  (mx-io/ndarray-iter [(get test-X-Y :X)]
                      {:label-name "linear_regression_label"
                       :label [(get test-X-Y :Y)]
                       :data-batch-size batch-size}))

(def train-iter
  (mx-io/ndarray-iter [(get train-X-Y :X)]
                      {:label-name "linear_regression_label"
                       :label [(get train-X-Y :Y)]
                       :data-batch-size batch-size}))

(defn train!
  [model-mod train-iter test-iter num-epoch]
  (-> model-mod
      (m/bind {:data-shapes (mx-io/provide-data train-iter)
               :label-shapes (mx-io/provide-label test-iter)})
      (m/fit {:train-data train-iter
              :eval-data test-iter
              ;; Training for `num-epochs`
              :num-epoch num-epoch
              :fit-params
              (m/fit-params
                {:batch-end-callback (callback/speedometer batch-size 100)
                 ;; Initializing weights with Xavier
                 :initializer (initializer/xavier)
                 ;; Choosing Optimizer Algorithm: SGD with lr = 0.01
                 :optimizer
                 (optimizer/sgd
                   {:learning-rate 0.01
                    :momentum 0.001
                    :lr-scheduler (lr-scheduler/factor-scheduler 3000 0.9)})
                 ;; Evaluation Metric
                 :eval-metric (eval-metric/mse)})})))


(comment

  ;; Defining parameters
  (def train-epoch 100)
  (def load-epoch 10000)
  (def prefix "mymodel")
  (def model-name prefix)
  (def data-names ["data"])
  (def label-names ["linear_regression_label"])

  ;; Wrapping the computation graph in a `module`
  (def model-mod
    (m/module (get-symbol)
              {:data-names data-names
               :label-names label-names}))

  ; (sym/list-arguments (m/symbol model-mod))
  ; (sym/list-outputs (m/symbol model-mod))

  ;; Training the model for `epoch`
  (train! model-mod train-iter test-iter train-epoch)

  ;; Persisting model to disk
  (m/save-checkpoint model-mod
                     {:epoch train-epoch
                      :prefix prefix
                      :save-opt-state true})

  ;; Loading model from disk
  (def loaded-model-mod
    (-> {:prefix prefix
         :load-optimizer-states true
         :epoch load-epoch
         :data-names data-names
         :label-names label-names}
        ;; Loading from checkpoint
        (m/load-checkpoint)
        ;; Binding label and data shapes
        (m/bind {:data-shapes (mx-io/provide-data train-iter)
                 :label-shapes (mx-io/provide-label test-iter)})))

  ;;; Performace evaluation

  ;; From `loaded-model-mod`
  (m/score loaded-model-mod
           {:eval-data test-iter
            :eval-metric (eval-metric/mse)})

  ;; From `model-mod`
  (m/score model-mod
           {:eval-data test-iter
            :eval-metric (eval-metric/mse)})

  (defn sanity-check
    [k model-mod test-iter test-X-Y]
    (let [score (m/score model-mod
                         {:eval-data test-iter
                          :eval-metric (eval-metric/mse)})
          predictions (-> model-mod
                          (m/predict {:eval-data test-iter})
                          (first)
                          (ndarray/slice 0 k)
                          (ndarray/->vec))
          ground-truth (-> test-X-Y
                           :Y
                           (ndarray/slice 0 k)
                           (ndarray/->vec))]
      (println "Score on Test Set: ")
      (println score)
      (println "\nPredictions: ")
      (println predictions)
      (println "\nGround Truth: ")
      (println ground-truth)))

  (sanity-check 20 model-mod test-iter test-X-Y)
  ;Score on Test Set:
  ;[mse 1.9759508]
  ;
  ;Predictions:
  ;[2.8569365 3.0594287 2.947115 2.841632 3.2931566 3.0408115 2.9448495 2.993783 2.841546 2.54163 2.89647 3.3247905 2.6727371 3.1266866 2.9160218 3.3124614 2.988596 3.0056143 3.1018448 3.086196]
  ;
  ;Ground Truth:
  ;[3.5 4.0 2.0 5.0 3.0 4.5 1.0 1.5 5.0 0.5 5.0 5.0 5.0 3.0 5.0 5.0 0.5 2.5 3.0 5.0]

  (sanity-check 20 loaded-model-mod test-iter test-X-Y)
  ;Score on Test Set:
  ;[mse 1.5319124]
  ;
  ;Predictions:
  ;[3.0426044 4.099924 2.0988526 3.3540404 3.194833 3.372664 2.748798 3.8139641 2.934378 1.3496954 2.891445 3.1200047 2.1070197 3.511092 3.3848426 3.853949 1.9194205 3.2366083 3.2725945 2.3981385]
  ;
  ;Ground Truth:
  ;[3.5 4.0 2.0 5.0 3.0 4.5 1.0 1.5 5.0 0.5 5.0 5.0 5.0 3.0 5.0 5.0 0.5 2.5 3.0 5.0]

;; Vizualisation
  (render-model! {:model-name model-name
                  :model-sym (get-symbol)
                  :input-data-shape [1 203]
                  :path "model_render"}))


ml.clj
(ns solrdemo-server.solr.ml
  (:require
   [clj-http.client :as http]
   [clojure.data.json :as json]
   [clojure.string :as string]
   [cortex.experiment.train :as train]
   [cortex.nn.layers :as layers]
   [cortex.nn.network :as network]
   [corona.client :as client]
   [corona.ltr :as ltr]
   [solrdemo-server.utils :as utils]
   [think.datatype.core :as dtype]))

(defn read-users
  "Parse user records from 'users-csv-resource-name' resource (.csv format)"
  []
  (let [parse-num (fn [v] (Float/parseFloat v))
        db-fields [:userId :gender :age :occupation :zip]
        users-fields-fns {:userId (fn [id] (Long/parseLong id))
                          :gender (fn [g] (case g "F" 0.0 "M" 1.0))
                          :age (fn [v] (when-not (= v "1") (Float/parseFloat v)))
                          :occupation (fn [v] (Float/parseFloat v))}
        entries (utils/read-csv-resource "solr/tmdb/data/users.csv" {:key-fn keyword :val-fns users-fields-fns})]
    (filter :age entries)))

#_(read-users)

#_(read-distinct-coll :spoken_languages :iso_639_1)

(defn build-features
  "Builds all project features for solr ltr.
  'store-name' - string name of store to specify for features (defaults to \"tmdb_features\")."
  [store-name movies]
  (mapcat identity
          [(ltr/gen-coll-features movies :genres store-name)
           (ltr/gen-coll-features movies :production_countries store-name)
           (ltr/gen-coll-features movies :spoken_languages store-name)
           ;; too many tags to use...

           [(ltr/gen-field-feature :popularity store-name)
            (ltr/gen-field-feature :vote_average store-name)
            (ltr/gen-field-feature :runtime store-name)
            (ltr/gen-field-feature :revenue store-name)
            (ltr/gen-field-feature :budget store-name)

            ;; user features
            (ltr/gen-external-value-feature :gender false store-name)
            (ltr/gen-external-value-feature :occupation false store-name)
            (ltr/gen-external-value-feature :age false store-name)]]))


#_(ltr/upload-features {:type :http :core :tmdb} (build-features "tmdb_features" [{}]))
#_(ltr/delete-feature-store {:type :http :core :tmdb})

(defn extract-mov-features
  "Extract features from solr
  executing HTTP GET request with query to 'sorl-core-url' (or http://localhost:8983/solr/tmdb).
  'store-name' - string name of store to extract features from (defaults to tmdb_features).
  Returns vector of maps with keys:
  :db_id - id of movie;
  :features - vector of features values for movie."
  [client-config store-name]
  (ltr/extract-features
   client-config
   {:q "*:*"
    :rows "10000"
    :sort "db_id asc"
    :fl "db_id"
    :store store-name}))



(defn read-ratings
  []
  (utils/read-csv-resource
   "solr/tmdb/data/ratings.csv"
   {:key-fn keyword
    :val-fns {:userId  (fn [id] (Long/parseLong id))
              :movieId (fn [id] (Long/parseLong id))
              :rating  (fn [f]  (Float/parseFloat f))}}))

#_(read-ratings)

(defn build-training-dataset
  "Builds dataset to use for train/test.
  Returns vector of maps with keys:
  :movieId, :userId, :score, :features
  where :features is vector of movie features concatenated with [user-gender user-age user-occupation]."
  []
  (let [users (read-users)
        mov-feats (extract-mov-features {:type :http :core :tmdb}
                                        "tmdb_features")]
    (->> (read-ratings)
         (map (fn [{:keys [userId movieId] :as rating}]
                (-> rating
                    (assoc :user  (first (filter #(= (:userId %) userId) users)))
                    (assoc :movie (first (filter #(= (:db_id %) (str movieId)) mov-feats))))))
         (remove (comp nil? :user))
         (remove (comp nil? :movie))
         (map (fn [{:keys [userId rating user movie]}]
                {:movieId  (Long/parseLong (:db_id movie))
                 :userId   userId
                 :score    [rating]
                 :features (vec
                            (concat
                             (drop-last 3 (:features movie))
                             [(:gender user) (:age user) (:occupation user)]))}))
         (doall))))

#_(build-training-dataset)

(defn normalize-training-dataset
  "Normalizes last 8 features (popularity vote_average runtime revenue budget gender occupation age) of dataset using MinMax norm.
  Performs simple undersampling for rates 3.0 4.0 5.0.
  Performs simple oversampling for rates 0.5 1.0 1.5.
  Returns map with keys
  :ds - normalized dataset with same structure as original;
  :mins - vector of minimums for last 8 features;
  :maxs - vector of maximums for last 8 features."
  [ds]
  (let [to-norm-count 8
        to-norm (->> ds
                     (map :features)
                     (map (partial take-last to-norm-count)))
        maxs (mapv (fn [index]
                     (->> to-norm
                          (map #(nth % index))
                          (reduce max 0)))
                   (range to-norm-count))
        mins (mapv (fn [index]
                     (->> to-norm
                          (map #(nth % index))
                          (reduce min Long/MAX_VALUE)))
                   (range to-norm-count))
        deltas (map - maxs mins)]
    {:ds   (->> ds
                (map #(update % :features ltr/do-to-last-n to-norm-count
                              (fn [index feat] (/ (- feat (nth mins index)) (nth deltas index)))))
                ;; dumb undersampling
                (remove (fn [{:keys [score]}] (and (= (first score) 3.0) (< (rand-int 10) 6))))
                (remove (fn [{:keys [score]}] (and (= (first score) 4.0) (< (rand-int 10) 7))))
                (remove (fn [{:keys [score]}] (and (= (first score) 5.0) (< (rand-int 10) 0))))
                ;; dumb oversampling
                (reduce (fn [ds entry]
                          (cond
                            (< (-> entry :score first) 1.0)
                            (vec (concat ds [entry entry entry entry]))
                            (< (-> entry :score first) 1.1)
                            (vec (concat ds [entry entry]))
                            (< (-> entry :score first) 1.6)
                            (vec (concat ds [entry entry entry entry]))
                            :else (conj ds entry)))
                        [])
                vec)
     :mins mins
     :maxs maxs}))

(defn train-nn
  "Builds classifier network as 4-layered perception, trains it using 'train-set' and 'test-set'.
  Returns map representation of trained network."
  [train-set test-set]
  (train/train-n
    (network/linear-network
      [(layers/input 203 1 1 :id :features)
       (layers/linear 512)
       (layers/relu)
       (layers/dropout 0.5)
       (layers/linear 128)
       (layers/relu)
       (layers/dropout 0.5)
       (layers/linear 16)
       (layers/relu)
       (layers/linear 1 :id :score)])
    train-set
    test-set
    :batch-size 200
    :network-filestem "ltr-goa-nn-new"
    :epoch-count 1000))

(defn get-layers
  "Extracts layers (weights and biases) of Cortex trained network 'cortex-trained-nn' in ready-for-json form.
  Returns vector of layers."
  [cortex-trained-nn]
  (let [{:keys [edges nodes buffers]} (-> cortex-trained-nn :compute-graph)]
    (->> edges
         (map (fn [[_ lay2k]]
                (let [lay2 (get nodes lay2k)
                      out-size (:output-size lay2)
                      in-size (-> lay2 :input-dimensions first :width)
                      w-buffer-id (-> lay2 :weights :buffer-id)
                      b-buffer-id (-> lay2 :bias :buffer-id)
                      w-bufs (get-in buffers [w-buffer-id :buffer])
                      b-buf (get-in buffers [b-buffer-id :buffer])]

                  (when (-> lay2 :type #{:linear})
                    {:matrix    (if (vector? w-bufs)
                                  (mapv (fn [buf]
                                          (mapv (fn [offset]
                                                  (dtype/get-value buf offset))
                                                (range in-size)))
                                        w-bufs)
                                  [(mapv (fn [offset]
                                           (dtype/get-value w-bufs offset))
                                         (range in-size))])

                     :bias       (mapv (fn [offset] (dtype/get-value b-buf offset))
                                       (range out-size))
                     :activation :relu}))))

         (remove nil?)
         vec
         (#(assoc-in % [(-> % count dec) :activation] :identity)))))

#_(get-layers trained-net)

#_(build-solr-ltr-model "tmdb_features" "ltrGoaModel" (get-layers trained-net) (:mins normed) (:maxs normed))

#_(ltr/delete-model! "ltrGoaModel")

(defn build-nn-model
  [trained-net normed movie-features]
  (ltr/build-solr-ltr-nn-model
   "tmdb_features"
   "ltrGoaModel" ;Gender Occupation Age
   (map #(select-keys % [:name]) movie-features)
   (get-layers trained-net)
   (:mins normed)
   (:maxs normed))) 
